packages/core/src/editor/BlockNoteEditor.ts

```diff
@@ -484,8 +485,28 @@ export class BlockNoteEditor<
     }
   }
 
-  dispatch(tr: Transaction) {
-    this._tiptapEditor.dispatch(tr);
+  /**
+   * Dispatch a transaction, but only if not in a "blocknote transact".
+   */
+  public dispatch(tr: Transaction) {
+    if (!this.activeTransaction) {
+      this._tiptapEditor.dispatch(tr);
+    }
+  }
+
+  /**
+   * Execute a function within a "blocknote transaction".
+   * All changes to the editor within the transaction will be grouped together, so that
+   * we can dispatch them as a single operation (thus creating only a single undo step)
+   */
+  public transact(func: () => void) {
+    this.activeTransaction = this._tiptapEditor.state.tr;
+    try {
+      func();
+      this._tiptapEditor.dispatch(this.activeTransaction);
+    } finally {
+      this.activeTransaction = undefined;
+    }
```

Comment:
Just for awareness, Tiptap already has a way to capture transactions without immediately dispatching them: https://github.com/ueberdosis/tiptap/blob/44cae5badc07e715bf6554e7042a9e4e926e57b7/packages/core/src/Editor.ts#L414-L429